pkg/controller/external.go

```diff
@@ -199,18 +208,39 @@ func (e *external) Observe(ctx context.Context, mg xpresource.Managed) (managed.
 		return managed.ExternalObservation{}, errors.Wrap(err, "cannot set observation")
 	}
 
+	// NOTE(lsviben) although the annotations were supposed to be set and the
+	// managed resource updated during the Create step, we are checking and
+	// updating the annotations here due to the fact that in most cases, the
+	// Create step is done asynchronously and the managed resource is not
+	// updated with the annotations. That is why below we are prioritizing the
+	// annotations update before anything else. We are setting lateInitialized
+	// to true so that the reconciler updates the managed resource. This
+	// behavior conflicts with management policies in which LateInitialize is
+	// turned off. To circumvent this, we are checking if the management policy
+	// does not contain LateInitialize and if it does not, we are updating the
+	// annotations manually.
 	annotationsUpdated, err := resource.SetCriticalAnnotations(tr, e.config, tfstate, string(res.State.GetPrivateRaw()))
 	if err != nil {
 		return managed.ExternalObservation{}, errors.Wrap(err, "cannot set critical annotations")
 	}
+	policyHasLateInit := policySet.HasAny(xpv1.ManagementActionLateInitialize, xpv1.ManagementActionAll)
+	if annotationsUpdated && !policyHasLateInit {
```

Comment:
So we are explicitly doing an update here for the sole case when:
- Management policies are enabled
- The MR's management policy lacks the `LateInitialize` policy

In this case, because in the current implementation we rely on the managed reconciler doing an update when the `managed.ExternalObservation.ResourceLateInitialized` is set to `True` (and I assume in the new crossplane-runtime, if late-initialization is disallowed, we will skip an update even though we return `ResourceLateInitialized: True` here), we have to update the resource. 

Is this correct? And is the only case we are doing this for is the case described above?

Reply:
Yeah, this is the case. 

Crossplane-runtime reconciler[ "expects"](https://github.com/crossplane/crossplane-runtime/blob/8e87b2c47ee705f88b1224bff7b3eb8d1a071760/pkg/reconciler/managed/reconciler.go#L1030) the critical annoatations to be updated after a Create. But, as Upjet mostly does [ApplyAsync](https://github.com/upbound/upjet/blob/accab08e93515620340fa2b3576fefbf14d70071/pkg/controller/external.go#L315), it doesnt set the Critical Annotations, instead it kinda misuses the LateInitialization mechanism to update the managed resource with the annotations. 

So yeah, this just makes sure that the annotations are still updated even if `LateInitialization` is off. So it should only affect resources in which:
- managementPolicies are enabled
- `LateInitialization` is off